### PR TITLE
ZEPPELIN-543: Storage initialization failure recovery

### PR DESCRIPTION
**Pull request by khalidhuseynov**
### What is this PR for?
Handling edge case failures when initializing notebook storage class(es).

### What type of PR is it?
Improvement

### Todos
* [x] - fix repo initialization
* [x] - add different config tests

### Is there a relevant Jira issue?
[ZEPPELIN-543](https://issues.apache.org/jira/browse/ZEPPELIN-543)

### How should this be tested?
There're various test cases (mostly included as unit tests).
One of the test scenarios:
1. in `conf/zeppelin-env.sh` add the following line
    ```
    export ZEPPELIN_NOTEBOOK_STORAGE="org.apache.zeppelin.notebook.repo.VFSNotebookRepo,org.apache.zeppelin.notebook.repo.DummyNotebookRepo"
    ```
2. Start Zeppelin
3. Previously Zeppelin would fail to start, now you would see that only first storage class is initialized and second just didn't initialize.

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by jongyoul** [comment]
@khalidhuseynov I've tested what you describe and saw the warning messages in a log and tests cases look nice to explain what you want to fix. LGTM.

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/NotebookRepoSyncInitializationTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    // create repo
+    NotebookRepoSync notebookRepoSync = new NotebookRepoSync(conf);
+    // check that second didn't initialize
+    LOG.info(" " + notebookRepoSync.getRepoCount());
+    assertEquals(notebookRepoSync.getRepoCount(), 1);
+    assertTrue(notebookRepoSync.getRepo(0) instanceof VFSNotebookRepo);
+  }
+  
+  @Test
+  public void InitUnsupporteNumberStoragesTest() throws IOException {
```
Comment (HeartSaVioR) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/NotebookRepoSyncInitializationTest.java]: nit: InitUnsupporte'd'NumberStoragesTest

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java [hunk 2]
```diff
@@ -50,28 +52,60 @@
...
    */
-  public NotebookRepoSync(ZeppelinConfiguration conf) throws Exception {
+  @SuppressWarnings("static-access")
+  public NotebookRepoSync(ZeppelinConfiguration conf) {
     config = conf;
 
     String allStorageClassNames = conf.getString(ConfVars.ZEPPELIN_NOTEBOOK_STORAGE).trim();
     if (allStorageClassNames.isEmpty()) {
-      throw new IOException("Empty ZEPPELIN_NOTEBOOK_STORAGE conf parameter");
+      allStorageClassNames = "org.apache.zeppelin.notebook.repo.VFSNotebookRepo";
```
Comment (HeartSaVioR) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: It would be better to use defaultStorage instead of string constant.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java [hunk 3]
```diff
@@ -50,28 +52,60 @@
...
+        repos.add((NotebookRepo) constructor.newInstance(conf));
+      } catch (ClassNotFoundException | NoSuchMethodException | SecurityException |
+          InstantiationException | IllegalAccessException | IllegalArgumentException |
+          InvocationTargetException e) {
+        LOG.warn("Failed to initialize {} notebook storage class {}", storageClassNames[i], e);
+      }
+    }
+    // couldn't initialize any storage, use default
+    if (getRepoCount() == 0) {
+      initializeDefaultStorage(conf);
```
Comment (HeartSaVioR) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/NotebookRepoSync.java]: nit: Personally I prefer to leave log when fail-back occurs. I'm curious others also prefer it. I don't have string opinion about this, so it shouldn't block this PR.

**Comment by HeartSaVioR** [comment]
I left some minor comments but code and unit tests looks great. LGTM

**Comment by HeartSaVioR** [comment]
After this PR, some repositories can't be loaded but we can't see what repositories are loaded via Web UI.
I think it may be better to have page to Zeppelin Web UI and let page shows loaded repositories.
(It could be filed to another JIRA issue.)
What do you think?

**Comment by jongyoul** [comment]
+1 @HeartSaVioR. It's better to display the warning

**Comment by khalidhuseynov** [comment]
@HeartSaVioR @jongyoul thanks for the feedback! Actually displaying warning to the user is a whole another issue that I came across few times recently, and I was thinking to address it soon. Definitely I'll create an issue and +1 for that.

**Comment by bzz** [comment]
+1 for better user notifications\warnings in the separate PR.
@khalidhuseynov could you please check if the method naming follows the project code conventions.

Shall we update and merge this one after that, if there is no discussion?

**Comment by khalidhuseynov** [comment]
@bzz thanks for the feedback! I updated naming for the methods I found starting with uppercase, let me know if there's anything else left. Otherwise it's good to go.

**Comment by bzz** [comment]
Thank you, looks great to me.
Merging if there is no more discussion.
